File: py-cplr/src/sdui_template_compiler/include_resolver.py
"""Component and fragment reference resolution from disk."""

import os
from dataclasses import dataclass
from typing import Any, Callable

from .yaml_loader import Yaml

JsonValue = Any


@dataclass(frozen=True)
class ResolvedInclude:
    file_path: str
    content: JsonValue


IncludeResolverFn = Callable[[str, str], ResolvedInclude]


class IncludeResolver:
    """Factory for disk-backed `.ref` resolvers."""

    @staticmethod
    def create(root_dir: str) -> IncludeResolverFn:
        """Creates a disk-backed resolver for component and fragment references.

        @param root_dir - SDUI root used as the base for reference paths
        @returns Resolver that returns an absolute path and parsed YAML content
        @throws When a referenced file is missing or is not JSON-compatible
        """

        def resolve_ref(ref_path: str, referrer_dir: str) -> ResolvedInclude:
            if ref_path.startswith("/"):
                file_path = os.path.abspath(
                    os.path.join(root_dir, "_components", f"{ref_path[1:]}.yaml")
                )
            else:
                file_path = os.path.abspath(os.path.join(referrer_dir, f"{ref_path}.yaml"))
            if not os.path.exists(file_path):
                raise ValueError(f"Unresolved .ref {ref_path}: {file_path}")

            return ResolvedInclude(file_path=file_path, content=Yaml.load(file_path))

        return resolve_ref

File: js-cplr/src/include-resolver.ts
import { existsSync } from 'node:fs';
import { resolve } from 'node:path';
import type { IncludeResolver as IncludeResolverFn } from './types.js';
import { Yaml } from './yaml.js';

export class IncludeResolver {
  /**
   * Creates a disk-backed resolver for component and fragment references.
   *
   * @param rootDir - SDUI root used as the base for reference paths
   * @returns Resolver that returns an absolute path and parsed YAML content
   * @throws When a referenced file is missing or is not JSON-compatible
   */
  static create(rootDir: string): IncludeResolverFn {
    return (refPath, referrerDir) => {
      const filePath = refPath.startsWith('/')
        ? resolve(rootDir, '_components', `${refPath.slice(1)}.yaml`)
        : resolve(referrerDir, `${refPath}.yaml`);
      if (!existsSync(filePath)) {
        throw new Error(`Unresolved .ref ${refPath}: ${filePath}`);
      }

      return { filePath, content: Yaml.load(filePath) };
    };
  }
}
